src/test/java/uk/gov/dwp/dataworks/DataKeyServiceApplicationTests.java

```diff
@@ -1,15 +1,126 @@
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.BDDMockito.given;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
 
 @RunWith(SpringRunner.class)
-@SpringBootTest
+@SpringBootTest()
+@ActiveProfiles("Test")
+@AutoConfigureMockMvc
 public class DataKeyServiceApplicationTests {
 
+    @Autowired
+    private CurrentKeyIdProvider currentKeyIdProvider;
+
+    @Autowired
+    private DataKeyGeneratorProvider dataKeyGeneratorProvider;
+
+    @Autowired
+    private DataKeyDecryptionProvider dataKeyDecryptionProvider;
+
+    @Autowired
+    private MockMvc mockMvc;
+
+    @Before
+    public void setup() {
+        Mockito.reset(currentKeyIdProvider, dataKeyGeneratorProvider, dataKeyDecryptionProvider);
+    }
+
     @Test
     public void contextLoads() {
     }
+
+    @Test
+    public void pingWorks() throws Exception {
+        mockMvc.perform(get("/ping"))
+                .andExpect(status().isOk())
+                .andExpect(content().string(""));
+    }
+
+    @Test
+    public void generateDataKeyRequest() throws Exception {
+        given(currentKeyIdProvider.getKeyId()).willReturn("myKeyId");
+
+        GenerateDataKeyResponse response = new GenerateDataKeyResponse("encryptionId", "plainKey", "cipherKey");
+
+        given(dataKeyGeneratorProvider.generateDataKey("myKeyId"))
+                .willReturn(response);
+
+        mockMvc.perform(get("/datakey"))
+                .andExpect(status().isCreated())
+                .andExpect(content().json(new ObjectMapper().writeValueAsString(response)));
+
+        // Must also allow a trailing slash
+        mockMvc.perform(get("/datakey/"))
+                .andExpect(status().isCreated())
+                .andExpect(content().json(new ObjectMapper().writeValueAsString(response)));
+    }
+
+    @Test
+    public void decryptRequest() throws Exception {
+        DecryptDataKeyResponse response = new DecryptDataKeyResponse("decryptKeyId", "iv", "plaintextDataKey");
+        given(dataKeyDecryptionProvider.decryptDataKey("encryptionKeyId", "cipher text"))
+                .willReturn(response);
+
+        mockMvc.perform(post("/datakey/actions/decrypt?keyId={keyId}", "encryptionKeyId").content("cipher text"))
+                .andExpect(status().isOk())
+                .andExpect(content().json(new ObjectMapper().writeValueAsString(response)));
+    }
+
+    @Test
+    public void currentKeyIdProviderThrowsError() throws Exception {
+        given(currentKeyIdProvider.getKeyId()).willThrow(new CurrentKeyIdFailure());
+
+        mockMvc.perform(get("/datakey"))
+                .andExpect(status().isServiceUnavailable());
```

Comment:
Might be worth asserting on the error content too as we are specifically throwing custom errors and custom responses.